### PR TITLE
Drop building in migrations

### PR DESCRIPTION
**Pull request by Arcticae**
TODO: 
 - [x] Add test for correctly resolving the contracts by name from the default build dir
 - [x] Make CI pass
 - [x] Update the docs

### CODE REVIEW

File: protostar/migrator/migrator_contract_identifier_resolver.py [hunk 1]
```diff
@@ -15,18 +30,15 @@ def __init__(
...
-            CompiledContractWriter(contract=contract_class, contract_name=contract_name)
-            .save(
-                output_dir=self._migrator_datetime_state.get_compilation_output_path()
-            )
-            .compiled_contract_path
+        return self._get_built_contract_by_name(contract_identifier)
+
+    def _get_built_contract_by_name(self, contract_name: str) -> Path:
+        compilation_output_dir = self._project_compiler.get_compilation_output_dir(
+            Path(BuildCommand.COMPILATION_OUTPUT_ARG.default)
```
Comment (kasperski95) [inline protostar/migrator/migrator_contract_identifier_resolver.py:40]: [Functionality]: This code will break if user will use name for the output directory. I think the migrate command should take extra argument instead.

**Review by kasperski95 — CHANGES_REQUESTED** [review]

File: protostar/migrator/migrator_contract_identifier_resolver.py (hunk 1 above)
Comment (Arcticae) [reply protostar/migrator/migrator_contract_identifier_resolver.py]: done

File: website/docs/tutorials/06-deploying/02-migrations/deploy-contract.md [hunk 2]
```diff
@@ -3,7 +3,7 @@
 ```python
 def deploy_contract(
     self,
-    contract_path: str,
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply website/docs/tutorials/06-deploying/02-migrations/deploy-contract.md:6]: done

File: protostar/commands/migrate_command.py [hunk 3]
```diff
@@ -100,6 +108,7 @@ async def migrate(
         output_dir_path: Optional[Path],
         migrator_config: MigratorExecutionEnvironment.Config,
         no_confirm: bool,
+        compiled_contracts_dir: Path,
```
Comment (kasperski95) [inline protostar/commands/migrate_command.py:111]: [Consistency] `output_dir_path` above uses different naming convention for the same thing

```suggestion
        output_dir_path: Optional[Path],
        migrator_config: MigratorExecutionEnvironment.Config,
        no_confirm: bool,
        compiled_contracts_dir_path: Path,
```

File: protostar/migrator/migrator_contract_identifier_resolver.py [hunk 4]
```diff
@@ -13,20 +27,19 @@ def __init__(
         self._project_compiler = project_compiler
         self._migrator_datetime_state = migrator_datetime_state
 
-    def resolve(self, contract_identifier: str) -> Path:
+    def resolve(self, contract_identifier: str, build_output_dir: Path) -> Path:
```
Comment (kasperski95) [inline protostar/migrator/migrator_contract_identifier_resolver.py:30]: [Design] `build_output_dir_path` can be injected directly to this class through constructor.

File: website/docs/tutorials/06-deploying/02-migrations/declare.md [hunk 5]
```diff
@@ -13,7 +13,7 @@ class DeclaredContract:
 
 Sends a declare transaction.
 
-- `contract_path_str` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, Protostar will compile the contract to the migration output directory.
+- `contract` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, protostar will look for it in the default build output directory (`build` in the project root).
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/declare.md:16]: [Documentation]
```suggestion
- `contract` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, Protostar will look for it in the [compiled-contracts-dir](LINK-TO-CLI-DOCS).
```

File: website/docs/tutorials/07-testing/02-cheatcodes/declare.md [hunk 6]
```diff
@@ -12,4 +12,5 @@ class DeclaredContract:
 ```
 Declares contract given a path relative to a Protostar project root.
 
-`config` is a keyword only argument kept for compatibility with the migration [declare cheatcode](../../06-deploying/02-migrations/declare.md). See related documentation for more information.
+- `contract` — A path to the **compiled** contract or a contract name. If you provide a contract name, Protostar will search for the compiled contract in the default build output (which is `build` in the root of the project) directory.
```
Comment (kasperski95) [inline website/docs/tutorials/07-testing/02-cheatcodes/declare.md:15]: [Documentation]
Testing cheatcodes don't use `ContractIdentifierResolver` hence this documentation is incorrect.

**Review by kasperski95 — CHANGES_REQUESTED** [review]

File: protostar/migrator/migrator_contract_identifier_resolver.py (hunk 4 above)
Comment (Arcticae) [reply protostar/migrator/migrator_contract_identifier_resolver.py]: done

File: website/docs/tutorials/06-deploying/02-migrations/declare.md (hunk 5 above)
Comment (Arcticae) [reply website/docs/tutorials/06-deploying/02-migrations/declare.md]: done

File: protostar/migrator/migrator_contract_identifier_resolver.py [hunk 7]
```diff
@@ -1,32 +1,48 @@
...
+from protostar.compiler import ProjectCompiler
 from protostar.migrator.migrator_datetime_state import MigratorDateTimeState
+from protostar.protostar_exception import ProtostarException
+
+
+class ContractIdentificationException(ProtostarException):
+    pass
+
+
+def checked_contract_path(contract_path: Path) -> Path:
```
Comment (kasperski95) [inline protostar/migrator/migrator_contract_identifier_resolver.py:12]:
```suggestion
def check_contract_path(contract_path: Path) -> Path:
```
[Consistency, Style] Methods and functions names usually start with a verb.

Comment (kasperski95) [reply protostar/migrator/migrator_contract_identifier_resolver.py:12]: [Style] This function could be below the class to follow [top-down vertical formatting](https://levelup.gitconnected.com/javascript-clean-code-vertical-formatting-d15097f5f30e).

@MaksymilianDemitraszek agreed to do that in the profiler:
https://github.com/software-mansion/protostar/pull/906#discussion_r988748014

File: website/docs/tutorials/06-deploying/02-migrations/declare.md [hunk 8]
```diff
@@ -13,7 +13,7 @@ class DeclaredContract:
 
 Sends a declare transaction.
 
-- `contract_path_str` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, Protostar will compile the contract to the migration output directory.
+- `contract` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, protostar will look for it in the [compiled-contracts-dir](../../../cli-reference.md#--compiled-contracts-dir-pathbuild).
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/declare.md:16]:
```suggestion
- `contract` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, Protostar will look for it in the [compiled-contracts-dir](../../../cli-reference.md#--compiled-contracts-dir-pathbuild).
```
[Consistency]

File: website/docs/tutorials/06-deploying/02-migrations/deploy-contract.md [hunk 9]
```diff
@@ -18,7 +18,7 @@ This cheatcode will be removed once [StarkNet drops support for deploy transacti
 :::
 
 Deploys a contract to the network.
-- `contract_path` — A path to the **compiled** contract or a contract name. If you provide a contract name, Protostar will compile the contract to the migration output directory.
+- `contract` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, protostar will look for it in the [compiled-contracts-dir](../../../cli-reference.md#--compiled-contracts-dir-pathbuild).
```
Comment (kasperski95) [inline website/docs/tutorials/06-deploying/02-migrations/deploy-contract.md:21]:
```suggestion
- `contract` — A path to the **compiled** contract or a [contract name](../../compiling#contract-name). If you provide a contract name, Protostar will look for it in the [compiled-contracts-dir](../../../cli-reference.md#--compiled-contracts-dir-pathbuild).
```
[Consistency]

**Review by kasperski95 — APPROVED** [review]
LGTM

File: protostar/migrator/migrator_contract_identifier_resolver.py (hunk 7 above)
Comment (Arcticae) [reply protostar/migrator/migrator_contract_identifier_resolver.py:12]: I wouldn't apply the rules from javascript here. Personally i like this style and unless we have some style guide we want to follow, i'm not really willing to change my mind :|

**Pull request merged**